docs/proposals/20250722-openstackclusteridentity.md

```diff
@@ -0,0 +1,213 @@
+# OpenStackClusterIdentity for Centralized Credential Management
+
+## Metadata
+
+- **Authors**: @bnallapeta
+- **Reviewers**: CAPO maintainers
+- **Status**: Proposed
+- **Creation Date**: 2025-07-22
+- **Last Updated**: 2025-07-22
+
+## Summary
+
+This proposal introduces `OpenStackClusterIdentity`, a cluster-scoped resource for centralized OpenStack credential management in CAPO. This enables multi-tenant environments to share credentials across namespaces while maintaining proper access controls, following patterns from AWS and Azure Cluster API providers.
+
+## Motivation
+
+### Goals
+- Enable centralized storage of OpenStack credentials in cluster-scoped resources
+- Provide fine-grained namespace access controls for credential usage
+- Maintain 100% backward compatibility with existing OpenStackCluster resources
+- Support manual, gradual migration without breaking existing deployments
+- Follow established patterns from other Cluster API providers (AWS, Azure)
+
+### Non-Goals
+- Automatic migration of existing deployments
+- Integration with external secret management systems
+- Breaking changes to existing API or functionality
+
+### User Stories
+
+#### Story 1: Platform Administrator
+As a platform administrator managing multiple tenant namespaces, I want to store OpenStack credentials centrally in a secure namespace (e.g., `capo-system`), control which tenant namespaces can use specific credentials, and rotate credentials in one place without updating every namespace.
+
+#### Story 2: Tenant User  
+As a tenant user in namespace `team-a`, I want to create OpenStack clusters using centrally managed credentials without managing OpenStack secrets in my namespace, with clear error messages if I don't have permission to use specific credentials.
+
+#### Story 3: Multi-Region Setup
+As an administrator managing clusters across multiple OpenStack regions, I want to create region-specific cluster identities with appropriate credentials and allow tenants to use different regional credentials based on their needs.
+
+### API Design
+
+#### New OpenStackClusterIdentity Resource (Cluster-scoped)
+
+```go
+type OpenStackClusterIdentity struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    Spec   OpenStackClusterIdentitySpec   `json:"spec,omitempty"`
+    Status OpenStackClusterIdentityStatus `json:"status,omitempty"`
+}
+
+type OpenStackClusterIdentitySpec struct {
+    // SecretRef references the secret containing OpenStack credentials
+    SecretRef OpenStackCredentialSecretReference `json:"secretRef"`
+
+    // AllowedNamespaces defines which namespaces can use this identity
+    // +optional
+    AllowedNamespaces []string `json:"allowedNamespaces,omitempty"`
```

Comment:
You could (and therefore probably should) drop this and just leave `NamespaceSelector`, btw: all namespaces have a `kubernetes.io/metadata.name` label containing their own name.

For example, `ValidatingWebhook` works this way: https://github.com/kubernetes/kubernetes/blob/e2eed70a296604f179171826d87172a9f2cf3add/pkg/apis/admissionregistration/types.go#L778-L822

Reply:
`allowedNamespaces` is simpler for the user if you just want specific namespaces. However, I'm just on the side of removal because:
- consistency with existing kubernetes APIs
- only having one way to do the thing

If we remove it I would definitely add a godoc comment to the namespace selector about `kubernetes.io/metadata.name`.